internal/provider/blob/blob.go

```diff
@@ -0,0 +1,105 @@
+package blob
+
+import (
+	"context"
+	"io"
+	"log"
+	"net/url"
+	"os"
+
+	"github.com/openshift-pipelines/tekton-caches/internal/tar"
+	"gocloud.dev/blob"
+
+	// Adding the driver for gcs.
+	_ "gocloud.dev/blob/gcsblob"
+	// Adding the driver for s3.
+	_ "gocloud.dev/blob/s3blob"
+	// If we want to add azure blob storage, we can use this import.
+	// _ "gocloud.dev/blob/azureblob" .
+)
+
+const (
+	cacheFile = "cache.tar.gz"
+)
+
+var (
+	queryParams string
+	openBucket  = func(ctx context.Context, urlString string) (*blob.Bucket, func() error, error) {
+		bucket, err := blob.OpenBucket(ctx, urlString+queryParams)
+		return bucket, bucket.Close, err
+	}
+)
+
+//nolint:gochecknoinits
+func init() {
+	queryParams = os.Getenv("BLOB_QUERY_PARAMS")
+}
+
+func Fetch(ctx context.Context, url url.URL, folder string) error {
+	log.Printf("Downloading cache from %s to %s", url.String(), folder)
+	file, err := os.CreateTemp("", cacheFile)
+	if err != nil {
+		log.Printf("error creating tar file: %s", err)
+		return err
+	}
+	defer os.Remove(file.Name())
+
+	bucket, clean, err := openBucket(ctx, url.String())
+	if err != nil {
+		log.Printf("error opening bucket: %s", err)
+		return err
+	}
+	defer clean() //nolint:errcheck
+
+	rc, err := bucket.NewReader(ctx, url.Path[1:], nil)
+	if err != nil {
+		log.Printf("error creating bucket reader: %s", err)
+		return err
+	}
+	defer rc.Close()
+
+	_, err = io.Copy(file, rc)
+	if err != nil {
+		log.Printf("error downloading cache: %s", err)
+		return err
+	}
+
+	if err := tar.Untar(ctx, file, folder); err != nil {
```

Comment:
i wonder if we can optimize the reading of the object and unar it to the same step without having to go by a temporary file locally.... this would greatly optimize it with a large cache, wdyt ?

Reply:
sounds good, can you create an issue for tracking?

Reply:
BTW, I change from io.Readall to io.copy. Won't `Readall` give OOM?

Reply:
Created an issue to track this.

Reply:
yeah io.Copy stream instead of reading everything at once in memory like ReadAll but that's beside the point that we can stream directly to gzip/upload to bucket.